modality-probe-capi/README.md

```diff
@@ -1,53 +1,172 @@
+management system for Rust is [Rustup](https://rustup.sh).
 
-## API
+### Building
+Once Rust is installed, build the C API using cargo:
 
-See the header file for documentation of available functions.
-
-## Building
-
-This library requires a [Rust](https://www.rust-lang.org/) toolchain.
-The recommended toolchain management system in Rust is [rustup](https://rustup.rs).
-
-Once rustup is installed, you can build for your local device with:
-
-```shell script
-cargo build
+```shell
+modality-probe/modality-probe-capi $ cargo build --release
 ```
 
-And the output artifact will be located in target/debug/libekotrace_capi.*
-
-Optimized libraries can be built with
-
-```shell script
-cargo build --release
+### Cross-platform Builds
+
+In order to build this library for platforms other than your host
+system, we recommend the use of cargo-xbuild. The default usage path
+for obtaining a new target's toolchain and building for likely looks
+something like the following. For more detailed information about
+cross-compilation for Rust, see
+[xbuild](https://github.com/rust-osdev/cargo-xbuild) and
+[rustc](https://github.com/japaric/rust-cross).
+
+```shell
+modality-probe/modality-probe-capi $ rustup update
+modality-probe/modality-probe-capi $ cargo install cargo-xbuild --force
+modality-probe/modality-probe-capi $ rustup target add thumbv7em-none-eabi
+modality-probe/modality-probe-capi $ rustc -Z unstable-options --print target-spec-json --target thumbv7em-none-eabi | tee thumbv7em-none-eabi.json
+modality-probe/modality-probe-capi $ cargo xbuild --target thumbv7em-none-eabi.json --release
 ```
 
-and are located at target/release/libekotrace_capi.*
-
-### Cross Platform Builds
+When using xbuild for cross-compilation, the output artifacts’
+locations follow the pattern target/<target>/<build-type>. Following
+with the example above, the artifacts would be placed at
+target/thumbv7em-none-eabi/release.
+
+## Usage
+
+The probe API consists of four behaviors: event recording, snapshot
+distribution and merging, “now”, and report generation.
+
+### Event Recording
+
+To record an event into a probe’s log, use one of the variants of
+`MODALITY_PROBE_RECORD`. In the example below, we’ve initialized a
+probe as a global called `probe_g`. You’ll also note that there are a
+few seemingly undefined symbols: `TWISTED` and `CONTROLLER`. These are
+given definitions by making use of Modality Probe’s CLI sub-command
+`header-gen` which you can read more about [here TODO(link)]. The
+short of it is that you’ll need to run the header-gen tool against
+your codebase before it can be compiled because Modality Probe
+actually generates the definitions to those symbols.
+
+```c
+#define DEFAULT_PROBE_SIZE = 7000;
+modality_probe * probe_g = MODALITY_PROBE_NULL_INITIALIZER;
+
+void twist(double x, double y, double z) {
+    int result = MODALITY_PROBE_RECORD(
+        probe_g,
+        TWISTED,
+        "tags=actuation",
```

Comment:
This is using the old tagging, now we have `MODALITY_TAGS(...)`: `MODALITY_TAGS("actuation")`